config.org
#+TITLE: Doom Configuration
#+AUTHOR: Sam Precious
#+EMAIL: samwdp@gmail.com
#+LANGUAGE: en
#+STARTUP: inlineimages
#+PROPERTY: header-args :tangle yes :cache yes :results silent :padline no

* Table Of Contents :toc:
- [[#configuration][Configuration]]
  - [[#global-optimisations][Global Optimisations]]
  - [[#personal-information][Personal Information]]
  - [[#doom-theming][Doom Theming]]
  - [[#tools][Tools]]
  - [[#languages][Languages]]
  - [[#key-remaps][Key Remaps]]
  - [[#include-files][Include Files]]

* Configuration
** Global Optimisations
*** Garbage collector
#+begin_src emacs-lisp
;; (defvar last-file-name-handler-alist file-name-handler-alist)
;; (setq gc-cons-threshold 402653184
;;       gc-cons-percentage 0.6
;;       file-name-handler-alist nil)
#+end_src
*** Unicode
#+BEGIN_SRC emacs-lisp
(setq inhibit-compacting-font-caches t)
#+END_SRC
** Personal Information
*** Name
#+BEGIN_SRC emacs-lisp
(setq user-full-name "Sam Precious")
#+END_SRC
*** Email Address
#+BEGIN_SRC emacs-lisp
(setq user-mail-address "samwdp@gmail.com")
#+end_src
*** Org Mode
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/Dropbox/org")
(setq org-log-done 'time)
#+END_SRC+END_SRC
** Doom Theming
*** Maximising
#+begin_src emacs-lisp
(toggle-frame-maximized)
#+end_src
*** Line Numbers
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type nil)
#+END_SRC
*** Fill column
#+begin_src emacs-lisp
(setq display-fill-column-indicator-column 80)
#+end_src
*** Font
**** Set fonts
#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Hasklug NF" :size 18 :weight 'light )
      doom-big-font (font-spec :family "Hasklug NF" :size 36)
      doom-variable-pitch-font (font-spec :family "NotoSans Nerd Font" :size 18))
(set-face-attribute 'fixed-pitch-serif (font-spec :family "NotoSans Nerd Font" :size 18))
#+END_SRC
**** Ligatures
#+begin_src emacs-lisp
;; you don't need to include all of them you can pick and mix
(plist-put! +ligatures-extra-symbols
  ;; org
  :name          "»"
  :src_block     "»"
  :src_block_end "«"
  :quote         nil
  :quote_end     nil
  ;; Functional
  :lambda        "λ"
  :def           "ƒ"
  :composition   "∘"
  :map           "↦"
  ;; Types
  :null          "∅"
  :true          "✓"
  :false         "✘"
  :int           nil
  :float         nil
  :str           nil
  :bool          nil
  :list          nil

  ;; Flow
  :not           "￢"
  :in            "∈"
  :not-in        "∉"
  :and           "∧"
  :or            "∨"
  :for           "∀"
  :some          "∃"
  :return        "⟼"
  :yield         "⟻"
  ;; Other
  :union         "⋃"
  :intersect     "∩"
  :diff          "∖"
  :tuple         "⨂"
  :pipe          "|" ;; FIXME: find a non-private char
  :dot           "•")  ;; you could also add your own if you want
#+end_src
*** Theme
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'my-doom-gruvbox)
#+END_SRC
*** Banner
#+BEGIN_SRC emacs-lisp
(setq fancy-splash-image (concat doom-private-dir "banners/default.png"))
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
#+END_SRC
*** Modeline
#+BEGIN_SRC emacs-lisp
(setq doom-modeline-buffer-file-name-style 'file-name)
(setq doom-modeline-buffer-encoding nil)
(setq doom-modeline-height 1)
(set-face-attribute 'mode-line nil :height 120)
(set-face-attribute 'mode-line-inactive nil :height 120)
#+END_SRC
*** Default Major mode for scratch buffer
Sets the default major mode of the scratch buffer to org-mode
#+begin_src emacs-lisp
(setq doom-scratch-initial-major-mode 'org-mode)
#+end_src
*** Default Directory
#+begin_src emacs-lisp
(setq default-directory "~")
#+end_src
*** Counsel
#+begin_src emacs-lisp
(setq company-idle-delay 0
      company-tooltip-idle-delay 0)
#+end_src
*** Centered Window
#+begin_src emacs-lisp
(setq centered-window-mode t)
#+end_src
** Tools
*** Debugger
#+begin_src emacs-lisp
(when IS-WINDOWS (setq dap-netcore-install-dir "C:/tools/"))
(setq +debugger--dap-alist `(((:lang csharp +lsp)     :after csharp-mode :require dap-netcore)
                             ((:lang cc +lsp)         :after ccls        :require (dap-lldb dap-gdb-lldb))
                             ((:lang python +lsp)     :after python      :require dap-python)
                             ((:lang rust +lsp)       :after rustic-mode :require (dap-lldb dap-cpptools))
                             ((:lang javascript +lsp) :after (js2-mode typescrit-mode) :require (dap-node dap-chrome))))
#+end_src
*** Treemacs
#+begin_src emacs-lisp
(when (featurep! :ui treemacs)
(setq treemacs-width 40
      treemacs-position 'right
      treemacs-indentation 2))
#+end_src

*** Neotree
#+begin_src emacs-lisp
(when (featurep! :ui neotree)
  (setq neo-theme 'ascii
        neo-window-position 'right
        neo-window-width 40))
#+end_src
*** Projectile
#+BEGIN_SRC emacs-lisp
(setq projectile-indexing-method 'native)
(when IS-WINDOWS
  (setq projectile-project-search-path '("W:\\foresolutions\\" "W:\\personal\\")))
#+END_SRC
*** Babel
Adds csharp and powershell languages to the org babel list
#+BEGIN_SRC emacs-lisp
(org-babel-do-load-languages
 'org-babel-load-languages
 '((csharp . t)
   (powershell . t)))
#+END_SRC
*** LSP
Sets some defaults for LSP mode doe more visuals
#+BEGIN_SRC emacs-lisp
(after! lsp-ui
  (setq lsp-ui-doc-enable t
        lsp-ui-doc-position 'top-right-corner
        lsp-ui-doc-alignment 'frame
        lsp-ui-doc-max-width  60
        lsp-ui-sideline-enable nil
        lsp-lens-enable t))
#+END_SRC
*** Ivy
#+begin_src emacs-lisp
(after! ivy-posframe
  (setq ivy-posframe-border-width 5))
#+end_src
*** Windows Specific
This is used in conjunction with shell mode to change the default shell to powershell core.
**** TODO This will probably change when the powershell core version updates
#+begin_src emacs-lisp
(when IS-WINDOWS
  (setq explicit-shell-file-name "C:/Program Files/PowerShell/7/pwsh.exe"))
#+end_src
** Languages
*** CSX
This is adding the ability to use csharp-mode with csharp script files
#+BEGIN_SRC emacs-lisp
(use-package! csharp-mode
  :mode (("\\.csx\\'" . csharp-mode)))
#+END_SRC
*** SQL
#+begin_src emacs-lisp
(use-package! lsp-mssql
  :config (add-hook! 'sql-mode-hook 'lsp))
#+end_src
** Key Remaps
*** Global Remaps
#+begin_src emacs-lisp
(map!
 :g "<f1>" #'cheat-sh
 :g "<f5>" #'dap-debug
 :g "<f11>" #'+lookup/references
 :g "<f12>" #'+lookup/definition
 :g "C-<f12>" #'+lookup/implementations
 :g "C-/" #'comment-line
 :leader
 :desc "Format buffer" "f o" #'+format/buffer
 :desc "Open Neotree" "o p" #'neotree-toggle)
#+end_src
*** LSP
#+begin_src emacs-lisp
(map! (:when (featurep! :tools lsp)
       :g "<f11>" #'+lookup/references
       :g "<f12>" #'+lookup/definition
       :g "C-<f12>" #'+lookup/implementations
       :g "<f3>" #'lsp-ui-doc-focus-frame
       :g "C-<f3>" #'lsp-ui-doc-unfocus-frame))
#+end_src
*** Csharp
#+begin_src emacs-lisp
(map! (:when (featurep! :lang csharp +dotnet)
        (:map csharp-mode-map
          :localleader
          :desc "Sharper" "s" #'sharper-main-transient)))
#+end_src
*** ELISP
#+begin_src emacs-lisp
(map! (:when (featurep! :lang emacs-lisp)
       (:map emacs-lisp-mode-map
        :g "<f3>" #'eval-buffer)))
#+end_src
*** Dap mode
#+begin_src emacs-lisp
(map! (:when (featurep! :tools debugger +lsp)
       (:map dap-mode-map
        :g "<f2>" #'dap-breakpoint-toggle
        :g "<f10>" #'dap-step-in
        :g "<f11>" #'dap-next
        :g "<f8>"  #'dap-continue)))
#+end_src
** Include Files
#+begin_src emacs-lisp
(defvar doom-user-private "doom-private")
(when (f-directory? (concat doom-private-dir doom-user-private))
  (mapc 'load (file-expand-wildcards (concat doom-private-dir doom-user-private "/*.el"))))
#+end_src


config.el
;; (defvar last-file-name-handler-alist file-name-handler-alist)
;; (setq gc-cons-threshold 402653184
;;       gc-cons-percentage 0.6
;;       file-name-handler-alist nil)
(setq inhibit-compacting-font-caches t)
(setq user-full-name "Sam Precious")
(setq user-mail-address "samwdp@gmail.com")
(toggle-frame-maximized)
(setq display-line-numbers-type nil)
(setq display-fill-column-indicator-column 80)
(setq doom-font (font-spec :family "Hasklug NF" :size 18 :weight 'light )
      doom-big-font (font-spec :family "Hasklug NF" :size 36)
      doom-variable-pitch-font (font-spec :family "NotoSans Nerd Font" :size 18))
(set-face-attribute 'fixed-pitch-serif (font-spec :family "NotoSans Nerd Font" :size 18))
;; you don't need to include all of them you can pick and mix
(plist-put! +ligatures-extra-symbols
  ;; org
  :name          "»"
  :src_block     "»"
  :src_block_end "«"
  :quote         nil
  :quote_end     nil
  ;; Functional
  :lambda        "λ"
  :def           "ƒ"
  :composition   "∘"
  :map           "↦"
  ;; Types
  :null          "∅"
  :true          "✓"
  :false         "✘"
  :int           nil
  :float         nil
  :str           nil
  :bool          nil
  :list          nil

  ;; Flow
  :not           "￢"
  :in            "∈"
  :not-in        "∉"
  :and           "∧"
  :or            "∨"
  :for           "∀"
  :some          "∃"
  :return        "⟼"
  :yield         "⟻"
  ;; Other
  :union         "⋃"
  :intersect     "∩"
  :diff          "∖"
  :tuple         "⨂"
  :pipe          "|" ;; FIXME: find a non-private char
  :dot           "•")  ;; you could also add your own if you want
(setq doom-theme 'my-doom-gruvbox)
(setq fancy-splash-image (concat doom-private-dir "banners/default.png"))
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)
(setq doom-modeline-buffer-file-name-style 'file-name)
(setq doom-modeline-buffer-encoding nil)
(setq doom-modeline-height 1)
(set-face-attribute 'mode-line nil :height 120)
(set-face-attribute 'mode-line-inactive nil :height 120)
(setq doom-scratch-initial-major-mode 'org-mode)
(setq default-directory "~")
(setq company-idle-delay 0
      company-tooltip-idle-delay 0)
(setq centered-window-mode t)
(when IS-WINDOWS (setq dap-netcore-install-dir "C:/tools/"))
(setq +debugger--dap-alist `(((:lang csharp +lsp)     :after csharp-mode :require dap-netcore)
                             ((:lang cc +lsp)         :after ccls        :require (dap-lldb dap-gdb-lldb))
                             ((:lang python +lsp)     :after python      :require dap-python)
                             ((:lang rust +lsp)       :after rustic-mode :require (dap-lldb dap-cpptools))
                             ((:lang javascript +lsp) :after (js2-mode typescrit-mode) :require (dap-node dap-chrome))))
(when (featurep! :ui treemacs)
(setq treemacs-width 40
      treemacs-position 'right
      treemacs-indentation 2))
(when (featurep! :ui neotree)
  (setq neo-theme 'ascii
        neo-window-position 'right
        neo-window-width 40))
(setq projectile-indexing-method 'native)
(when IS-WINDOWS
  (setq projectile-project-search-path '("W:\\foresolutions\\" "W:\\personal\\")))
(org-babel-do-load-languages
 'org-babel-load-languages
 '((csharp . t)
   (powershell . t)))
(after! lsp-ui
  (setq lsp-ui-doc-enable t
        lsp-ui-doc-position 'top-right-corner
        lsp-ui-doc-alignment 'frame
        lsp-ui-doc-max-width  60
        lsp-ui-sideline-enable nil
        lsp-lens-enable t))
(after! ivy-posframe
  (setq ivy-posframe-border-width 5))
(when IS-WINDOWS
  (setq explicit-shell-file-name "C:/Program Files/PowerShell/7/pwsh.exe"))
(use-package! csharp-mode
  :mode (("\\.csx\\'" . csharp-mode)))
(use-package! lsp-mssql
  :config (add-hook! 'sql-mode-hook 'lsp))
(map!
 :g "<f1>" #'cheat-sh
 :g "<f5>" #'dap-debug
 :g "<f11>" #'+lookup/references
 :g "<f12>" #'+lookup/definition
 :g "C-<f12>" #'+lookup/implementations
 :g "C-/" #'comment-line
 :leader
 :desc "Format buffer" "f o" #'+format/buffer
 :desc "Open Neotree" "o p" #'neotree-toggle)
(map! (:when (featurep! :tools lsp)
       :g "<f11>" #'+lookup/references
       :g "<f12>" #'+lookup/definition
       :g "C-<f12>" #'+lookup/implementations
       :g "<f3>" #'lsp-ui-doc-focus-frame
       :g "C-<f3>" #'lsp-ui-doc-unfocus-frame))
(map! (:when (featurep! :lang csharp +dotnet)
        (:map csharp-mode-map
          :localleader
          :desc "Sharper" "s" #'sharper-main-transient)))
(map! (:when (featurep! :lang emacs-lisp)
       (:map emacs-lisp-mode-map
        :g "<f3>" #'eval-buffer)))
(map! (:when (featurep! :tools debugger +lsp)
       (:map dap-mode-map
        :g "<f2>" #'dap-breakpoint-toggle
        :g "<f10>" #'dap-step-in
        :g "<f11>" #'dap-next
        :g "<f8>"  #'dap-continue)))
(defvar doom-user-private "doom-private")
(when (f-directory? (concat doom-private-dir doom-user-private))
  (mapc 'load (file-expand-wildcards (concat doom-private-dir doom-user-private "/*.el"))))
